### PR TITLE
Snow level

### PR DESCRIPTION
**Pull request by Ph4i1ur3**
Adds support for Snow to the relevent `.level` tags and mechs. 
Also contains general file and meta cleanup, for maintainability.

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 1]
```diff
@@ -50,41 +51,67 @@ public static void registerTags() {
         // @returns ElementTag(Number)
         // @group properties
         // @description
-        // Returns the maximum level for a levelable material (like water, lava, and Cauldrons), a cake, or a beehive.
+        // Returns the maximum level for a valid material.
+        // Valid materials include water, lava, cauldrons, cake, beehives, and snow.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: Please change this back to the way it was. `a levelable material (like water, lava, and Cauldrons)` is NOT a definitive / complete list.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 2]
```diff
@@ -50,41 +51,67 @@ public static void registerTags() {
...
             return new ElementTag(material.getMax());
         });
 
+        // <--[tag]
+        // @attribute <MaterialTag.minimum_level>
+        // @returns ElementTag(Number)
+        // @group properties
+        // @description
+        // Returns the minimum level for a valid material.
+        // Valid materials include water, lava, cauldrons, cake, beehives, and snow.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: This meta should clarify that there's only a non-zero value for snow (if it even is actually non-zero for snow?)

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 3]
```diff
@@ -50,41 +51,67 @@ public static void registerTags() {
...
-        // Returns the current level for a levelable material (like water, lava, and Cauldrons), a cake, or a beehive.
+        // Returns the current level for a valid material.
+        // Valid materials include water, lava, cauldrons, cake, beehives, and snow.
         // -->
         PropertyParser.<MaterialLevel>registerTag("level", (attribute, material) -> {
             return new ElementTag(material.getCurrent());
         });
     }
 
+    // Levelled Interface
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: Remove these pointless comments

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 4]
```diff
@@ -50,41 +51,67 @@ public static void registerTags() {
...
-    public Cake getCake() {
-        return (Cake) material.getModernData().data;
+    public Snow getSnow() {
+        return (Snow) material.getModernData().data;
     }
 
+    // Beehive Interface
+    public boolean isHive() {
+            return (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_15) && material.getModernData().data instanceof Beehive);
+        }
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: something went funky with the formatting here

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 5]
```diff
@@ -50,41 +51,59 @@ public static void registerTags() {
         // @returns ElementTag(Number)
         // @group properties
         // @description
-        // Returns the maximum level for a levelable material (like water, lava, and Cauldrons), a cake, or a beehive.
+        // Returns the maximum level for a level-able material (like water, lava, and cauldrons).
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: Please either use `levelable` as was written, or `Levelled` to match the spigot API name.

Comment (mcmonkey4eva) [reply plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: Also please restore the list of items back to the way it was.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 6]
```diff
@@ -50,41 +51,59 @@ public static void registerTags() {
...
             return new ElementTag(material.getMax());
         });
 
+        // <--[tag]
+        // @attribute <MaterialTag.minimum_level>
+        // @returns ElementTag(Number)
+        // @group properties
+        // @description
+        // Returns the minimum level for a level-able material (like water, lava, and cauldrons).
+        // This will return 0 for all level-able materials aside from snow.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: Snow is not a Levelled.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 7]
```diff
@@ -50,41 +51,59 @@ public static void registerTags() {
...
+        return material.getModernData().data instanceof Snow;
     }
 
-    public Cake getCake() {
-        return (Cake) material.getModernData().data;
+    public Snow getSnow() {
+        return (Snow) material.getModernData().data;
     }
 
+    public boolean isHive() { return (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_15) && material.getModernData().data instanceof Beehive); }
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: wtf happened here?

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 8]
```diff
@@ -96,11 +115,15 @@ public int getMaxHoneyLevel() {
     public void setHoneyLevel(int level) {
         ((Beehive) material.getModernData().data).setHoneyLevel(level);
     }
-    
+
+    // Functions
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: Remove these pointless comments.

**Comment by Ph4i1ur3** [comment]
I *think* i did it,

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 9]
```diff
@@ -111,17 +136,31 @@ public int getMax() {
         if (isCake()) {
             return getCake().getMaximumBites();
         }
+        if (isSnow()) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: This should be an `else`

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java [hunk 10]
```diff
@@ -50,39 +51,60 @@ public static void registerTags() {
...
 
-    public Cake getCake() {
-        return (Cake) material.getModernData().data;
+    public Snow getSnow() {
+        return (Snow) material.getModernData().data;
+    }
+
+    public boolean isHive() {
+        return (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_15)
+                && material.getModernData().data instanceof Beehive);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/material/MaterialLevel.java]: This should be one line, not two.

**Comment by mcmonkey4eva** [comment]
:D

**Pull request merged**